### ISSUE TITLE
Error in Multiple Jobs Scheduling

### ISSUE THREAD

**Issue by gs-spadmanabhan** [EasyBatch/easybatch-framework#65]
Scenario:

Executing BatchSchedulerTest and in that especially testMultipleJobsScheduling() test method. Instead of current engine instances please use the below, 

Engine engine1 = new EngineBuilder().reader(new StringRecordReader("foo\nbar\n")).build();
Engine engine2 = new EngineBuilder().reader(new StringRecordReader("foo\nbar\nbarfoo\n")).build();

Even though the test case is supposed to run 2 jobs, it runs only the engine2 job somehow the scheduler is picking up only the engine2 instance and skipping the engine1 instance.

**Comment by benas** [comment]
Hi Sunand,

Thank you for pointing out this issue. In fact, the scheduler is somehow ignoring the first engine. I'm able to reproduce the bug with this failing test using version 3.0.2-SNAPSHOT:

```java
    @Mock
    private RecordReader recordReader1, recordReader2;
	
	@org.junit.Test
    public void testMultipleJobsScheduling() throws Exception {

        Mockito.when(recordReader1.hasNextRecord()).thenReturn(false);
        Mockito.when(recordReader2.hasNextRecord()).thenReturn(false);
        
        Engine engine1 = new EngineBuilder().named("engine1").reader(recordReader1).build();
        Engine engine2 = new EngineBuilder().named("engine2").reader(recordReader2).build();

        BatchScheduler batchScheduler1 = new BatchScheduler(engine1);
        batchScheduler1.scheduleAtWithInterval(new Date(), 1);
        batchScheduler1.start();

        BatchScheduler batchScheduler2 = new BatchScheduler(engine2);
        batchScheduler2.scheduleAtWithInterval(new Date(), 1);
        batchScheduler2.start();

        assertThat(batchScheduler1.isStarted()).isTrue();
        assertThat(batchScheduler2.isStarted()).isTrue();

        batchScheduler1.stop();
        batchScheduler2.stop();

        Mockito.verify(recordReader1).hasNextRecord();
        Mockito.verify(recordReader2).hasNextRecord();

    }
```

I get the following log output:

```
[Mon Jun 08 16:56:12 CEST 2015] INFO: Building a scheduler for job batch-job-74e6a7b7-8139-4bcb-b22f-fd9baf635a9a to start at 08/06/15 16:56 and every 1 minute(s)
[Mon Jun 08 16:56:12 CEST 2015] INFO: Starting the scheduler for job batch-job-74e6a7b7-8139-4bcb-b22f-fd9baf635a9a
[Mon Jun 08 16:56:12 CEST 2015] INFO: Building a scheduler for job batch-job-83cfb33a-9058-4fce-ae3c-7255a15755b6 to start at 08/06/15 16:56 and every 1 minute(s)
[Mon Jun 08 16:56:12 CEST 2015] INFO: Starting the scheduler for job batch-job-83cfb33a-9058-4fce-ae3c-7255a15755b6
[Mon Jun 08 16:56:12 CEST 2015] INFO: Initializing the engine
[Mon Jun 08 16:56:12 CEST 2015] INFO: Initializing the engine
[Mon Jun 08 16:56:12 CEST 2015] INFO: Engine name: engine2
[Mon Jun 08 16:56:12 CEST 2015] INFO: Engine name: engine2
[Mon Jun 08 16:56:12 CEST 2015] INFO: Execution id: 83cfb33a-9058-4fce-ae3c-7255a15755b6
[Mon Jun 08 16:56:12 CEST 2015] INFO: Execution id: 83cfb33a-9058-4fce-ae3c-7255a15755b6
[Mon Jun 08 16:56:12 CEST 2015] INFO: Strict mode: false
[Mon Jun 08 16:56:12 CEST 2015] INFO: Strict mode: false
[Mon Jun 08 16:56:12 CEST 2015] INFO: Data source: N/A
[Mon Jun 08 16:56:12 CEST 2015] INFO: Data source: N/A
[Mon Jun 08 16:56:12 CEST 2015] INFO: The engine is running
[Mon Jun 08 16:56:12 CEST 2015] INFO: The engine is running
[Mon Jun 08 16:56:12 CEST 2015] INFO: Shutting down the engine
[Mon Jun 08 16:56:12 CEST 2015] INFO: Shutting down the engine
[Mon Jun 08 16:56:12 CEST 2015] INFO: Stopping the scheduler for job batch-job-74e6a7b7-8139-4bcb-b22f-fd9baf635a9a
[Mon Jun 08 16:56:12 CEST 2015] INFO: Stopping the scheduler for job batch-job-83cfb33a-9058-4fce-ae3c-7255a15755b6

Wanted but not invoked:
recordReader1.hasNextRecord();
-> at org.easybatch.integration.quartz.BatchSchedulerTest.testMultipleJobsScheduling(BatchSchedulerTest.java:108)
Actually, there were zero interactions with this mock.
```

First, there is no interaction with the reader of the first engine, and second, it is clear from the output that engine1 has not been called at all:

```
[Mon Jun 08 16:56:12 CEST 2015] INFO: Engine name: engine2
[Mon Jun 08 16:56:12 CEST 2015] INFO: Engine name: engine2
[Mon Jun 08 16:56:12 CEST 2015] INFO: Execution id: 83cfb33a-9058-4fce-ae3c-7255a15755b6
[Mon Jun 08 16:56:12 CEST 2015] INFO: Execution id: 83cfb33a-9058-4fce-ae3c-7255a15755b6
```

I'll fix the issue asap and keep you informed.

Kind regards
Mahmoud

**Comment by gs-spadmanabhan** [comment]
Hi,
The problem is with the Scheduler. It is only taking the last job Instance and executing it (skips the rest). After breaking my head I see the only way to solve this issue is by using JobDataMap of quartz scheduler in which we can pass the batch instance. I encountered this problem while writing a similar scheduler for easy-rules. I have solved this problem locally in easy-rules and happy to give you a Pull Request. Let me know if you find another way.

Thanks!
Sunand

**Comment by benas** [comment]
Hi,

In fact the problem is with the Scheduler. The `StdSchedulerFactory` provided by quartz seems to be a singleton and hence has only one scheduler that is overridden with the last engine when called multiple times. Here is a failing test:

```java
@Test
public void testDefaultQuartzStdSchedulerFactory() throws Exception {
    SchedulerFactory schedulerFactory1 = new StdSchedulerFactory();
    SchedulerFactory schedulerFactory2 = new StdSchedulerFactory();
    Scheduler scheduler1 = schedulerFactory1.getScheduler();
    Scheduler scheduler2 = schedulerFactory2.getScheduler();
    assertThat(scheduler1).isNotEqualTo(scheduler2);
}
```
Even though the API allows you to create 2 instances of the `StdSchedulerFactory`, you get the same scheduler instance! I was expecting getting a different scheduler each time I create a new `StdSchedulerFactory`.

IMO, quartz should not provide a public default constructor for such a singleton because this is error prone.

Now, according to [quartz doc](http://quartz-scheduler.org/documentation/quartz-2.x/configuration/ConfigMain) (see `org.quartz.scheduler.instanceName` description), it is possible to create multiple scheduler instances within the same app, we need just to give them different names. So here is how to do it:

```java
@Test
public void testQuartzStdSchedulerFactoryWithUniqueName() throws Exception {
    // load quartz default configuration
    Properties properties1 = new Properties();
    properties1.load(this.getClass().getResourceAsStream("/org/quartz/quartz.properties"));
    // override the scheduler name property to make it unique
    properties1.setProperty("org.quartz.scheduler.instanceName", "1");
    SchedulerFactory schedulerFactory1 = new StdSchedulerFactory(properties1);
    Scheduler scheduler1 = schedulerFactory1.getScheduler();

    // load quartz default configuration
    Properties properties2 = new Properties();
    properties2.load(this.getClass().getResourceAsStream("/org/quartz/quartz.properties"));
    // override the scheduler name property to make it unique
    properties2.setProperty("org.quartz.scheduler.instanceName", "2");
    SchedulerFactory schedulerFactory2 = new StdSchedulerFactory(properties2);
    Scheduler scheduler2 = schedulerFactory2.getScheduler();

    assertThat(scheduler1).isNotEqualTo(scheduler2);
}
```
This test is passing since each scheduler has its own name.
In Easy Batch , the scheduler name will be the engine execution id which is unique.

I've fixed this issue and deployed version 3.0.2-SNAPSHOT to maven central snapshot repository.

Please give it a try and keep me informed if it is ok for you. Many thanks upfront :-)

Kind regards,
Mahmoud

**Comment by gs-spadmanabhan** [comment]
Hi Mahmoud,

This fix is great, thanks for the update and explanation. Tryout jobDataMap as well, the reason is in most cases we will have a single scheduler running infinitely and we will be scheduling jobs at a given time. 

I learned something new, thanks for that.

- Sunand.

**Comment by benas** [comment]
Hi Sunand,

My fix is actually not that great.. Your fix is better! If we create multiple schedulers, quartz will create multiple thread pools behind the scene which is very expensive. With JobDataMap, we actually need only one scheduler which is by far more efficient. I've [updated](https://github.com/EasyBatch/easybatch-framework/commit/0d8aa72ad40e20494452ce5ffc9b50f2c5f0e251) the fix with your suggestion and [added](d99b9a97ad004df8c22bf32f3b2bb716cadf3ff0) you to the awesome contributors list :-)

Thank you!
Best regards
Mahmoud

**Comment by gs-spadmanabhan** [comment]
Cool :+1:
